pipeline/data_store.py

```diff
@@ -0,0 +1,773 @@
+import sqlalchemy as sa
+
+from pipeline.utils import get_git_hash, get_latest_provenance
+
+from models.base import SmartSession
+from models.provenance import CodeHash, CodeVersion, Provenance
+from models.exposure import Exposure
+from models.image import Image
+from models.source_list import SourceList
+from models.world_coordinates import WorldCoordinates
+from models.zero_point import ZeroPoint
+from models.cutouts import Cutouts
+from models.measurements import Measurements
+
+
+UPSTREAM_NAMES = {
+    'preprocessing': [],
+    'extraction': ['preprocessing'],
+    'astrometry': ['extraction'],
+    'calibration': ['extraction', 'astrometry'],
+    'subtraction': ['preprocessing', 'extraction', 'astrometry', 'calibration'],
+    'detection': ['subtraction'],
+    'cutting': ['detection'],
+    'measurement': ['detection', 'calibration'],
+}
+
+UPSTREAM_OBJECTS = {
+    'preprocessing': 'image',
+    'coaddition': 'image',
+    'extraction': 'sources',
+    'astrometry': 'wcs',
+    'calibration': 'zp',
+    'subtraction': 'sub_image',
+    'detection': 'detections',
+    'cutting': 'cutouts',
+    'measurement': 'measurements',
+}
+
+
+class DataStore:
```

Comment:
A question about how this is going to be used.

Do you visualize a single DataStore object for each run through the pipeline?

If so, the structure if it may not be flexible enough.  For instance, we're always going to have at least two images, the science image, and the reference image.  Is the idea that you don't cache the reference image in the datastore?  If so, where would the lazy loading of the reference data be handled?   I can also think of a reason why we might want to read up to four catalogs during the processing.  There's the catalog extracted from the search image, the catalog extracted from the reference image, the catalog used for matching for astrometry, and the catalog used for matching for photometry.  Where is the lazy-loading handled for the last three catalogs?

And I doubt I've thought of everything right here; there may well come other reasons why we have to load ancillary images or catalogs.  For instance, if we decide to get more sophisticated in dealing with color terms, then we might have to deal with mutpiple images at once.  While I don't think we should worry about that as we're first building the pipeline (as that adds all kinds of complication), having a datastore structure set up to be able to handle something like that would be nice (so that we don't have to rewrite everything that uses the datastore later as we realize we need to add something like that). 

We could start adding individual things for each thing like this we think of.  However, in some cases the code is the same, e.g. the catalog for the search and reference images are both produced by our code and will be loaded in the same (or almost the same) way.  A more general datastore for catalogs would make sense then, which could then be keyed here rather than defining as object attributes all of the ones we end up needing in any branch of the pipeline.

Reply:
Yes, the datastore is used to parse a few inputs (like exp/ccd id) and then can lazy load and cache all the information you need for the analysis. 
There's a `ref_image` and `sub_image` in the datastore. But of course you may need more data to be added. I don't future proof my code (that's impossible) so if you need to add things, you'd just add them later. 

Also, because I am using "getter" functions, that have the ability to lazy load from DB, we can easily expand some of the functionality without altering the interface (for example, if we end up with more catalogs). 

Right now I have "sources" which is the catalog from the science image (I am avoiding the word catalog because I think that is a good name for the cross-matching catalogs like Gaia, instead I'm using source list). I also have "detections" which is the exact same thing only on the subtracted image. 

I don't know why we'd need a separate source list for astrometry and photometry? 

The reference image needs a list of sources so we can compare to the new one and swarp them into the same reference, I guess? 

If so, that source list should come with the ref image, not re-produced by us each time.

Reply:
The list I was talking about for astrometry and photometry are the catalogs with the calibrated information we're matching to.

Reply:
A practical issue we're going to have to confront: we can't have too many processes connecting to external services at the same time, or we're going to start getting failures.

For example, the gaia servers.  Right now with lensgrinder, when I'm running 60 chips, a single process downloads the astrometric data for all of them all at once.  That's a single connection to the gaia server.  External servers are going to start kicking us off if we make too many simultaneous connections, which will happen if each process runs independently and tries to get the astrometric data at the same time.

I do similar things for photometric data, and for downloading survey sources for building references.

One solution is local caching (which I also do) -- save the astrometric data for a field, and we never have to go to gaia again for it.  However, that doesn't help when we're doing a new field.

I don't know if this can be solved if we want to deal with each chip as a completely independent process, as the current design suggests.  Lensgrinder does have some exposure-level processes for things like this.